src/controllers/plugins/api/JSONStringify.cpp

```diff
@@ -0,0 +1,252 @@
+
+void stringifyArray(lua_State *L, auto &writer, uint16_t depth, size_t length)
+{
+    lua_checkstack(L, 2);  // value + potential error
+
+    writer.StartArray();
+
+    for (size_t i = 0; i < length; i++)
+    {
+        lua_rawgeti(L, -1, static_cast<lua_Integer>(i) + 1);
+        stringifyValue(L, writer, depth);
+        lua_pop(L, 1);  // pop value
+    }
+
+    writer.EndArray();
+}
+
+void stringifyObject(lua_State *L, auto &writer, uint16_t depth)
+{
+    lua_checkstack(L, 3);  // key + value + potential error
+
+    writer.StartObject();
+
+    size_t items = 0;
+    lua_pushnil(L);
+    while (lua_next(L, -2) != 0)
+    {
+        items++;
+        if (items > ENCODE_MAX_TABLE_LENGTH)
+        {
+            luaL_error(L, "Too many items in table");
+            std::terminate();
+        }
+        // Stack:
+        // [-3] table
+        // [-2] key
+        // [-1] value
+        auto keyType = lua_type(L, -2);
+        if (keyType != LUA_TSTRING)
+        {
+            luaL_error(L, "Object key was not a string, got %s",
+                       lua_typename(L, keyType));
+            std::terminate();
+        }
+        size_t len = 0;
+        const char *str = lua_tolstring(L, -2, &len);
+        writer.Key(str, len, /*copy=*/true);
+
+        stringifyValue(L, writer, depth);
+        lua_pop(L, 1);  // pop value
+    }
+
+    writer.EndObject();
+}
+
+void stringifyValue(lua_State *L, auto &writer, uint16_t depth)
+{
+    auto typeId = lua_type(L, -1);
+    switch (typeId)
+    {
+        case LUA_TNIL:
+            writer.Null();
+            break;
+        case LUA_TBOOLEAN:
+            writer.Bool(lua_toboolean(L, -1) != 0);
+            break;
+        case LUA_TNUMBER: {
+            if (lua_isinteger(L, -1))
+            {
+                writer.Int64(lua_tointeger(L, -1));
+            }
+            else
+            {
+                writer.Double(lua_tonumber(L, -1));
+            }
+        }
+        break;
+        case LUA_TSTRING: {
+            size_t len = 0;
+            const char *str = lua_tolstring(L, -1, &len);
+            writer.String(str, len, /*copy=*/true);
+        }
+        break;
+        case LUA_TTABLE: {
+            if (depth >= ENCODE_MAX_DEPTH)
+            {
+                luaL_error(L, "Too deep");
+                std::terminate();
+            }
+
+            auto arraySize = inferArraySize(L);
+            if (arraySize)
+            {
+                stringifyArray(L, writer, depth + 1, *arraySize);
+            }
+            else
+            {
+                stringifyObject(L, writer, depth + 1);
+            }
+        }
+        break;
+        case LUA_TLIGHTUSERDATA: {
+            auto *ptr = lua_touserdata(L, -1);
+            if (ptr == nullptr)
+            {
+                writer.Null();
+            }
+            else
+            {
+                luaL_error(L, "Non-null lightuserdata not supported");
```

Comment:
This error text does not make sense without knowing implementation details of this module.